internal/graph/check.go

```diff
@@ -984,6 +991,491 @@ func trySendUsersetsAndDeleteFromMap(ctx context.Context, usersetsMap usersetsMa
+							// It probably is +1.  However, we have no control on whether
+							// the userset member lookup for the first level has started DS query.
+							DatastoreQueryCount: req.GetRequestMetadata().DatastoreQueryCount + 2,
+						},
+					}, nil, nil, nil
+				}
+			} else {
+				found, err := processUsersetMessage(userToUsersetMessage, usersetFromUser, usersetFromObject)
+				if err != nil {
+					return nil, nil, nil, err
+				}
+				if found {
+					return &ResolveCheckResponse{
+						Allowed: true,
+						ResolutionMetadata: &ResolveCheckResponseMetadata{
+							DatastoreQueryCount: req.GetRequestMetadata().DatastoreQueryCount + 2,
+						},
+					}, nil, nil, nil
+				}
+				//  check to see if there is a direct assignment
+				if req.GetTupleKey().GetObject() == userToUsersetMessage.userset {
+					return &ResolveCheckResponse{
+						Allowed: true,
+						ResolutionMetadata: &ResolveCheckResponseMetadata{
+							DatastoreQueryCount: req.GetRequestMetadata().DatastoreQueryCount + 2,
+						},
+					}, nil, nil, nil
+				}
+			}
+		case objectToUsersetMessage, ok := <-objectToUsersetMessageChan:
+			if !ok {
+				objectToUsersetDone = true
+				// note that if channel is closed and usersetFromObject is empty, it is still possible
+				// for allowed:true IF there is a direct assignment. Therefore, we will need to wait
+				// until userset has finished processing.
+			} else {
+				found, err := processUsersetMessage(objectToUsersetMessage, usersetFromObject, usersetFromUser)
+				if err != nil {
+					return nil, nil, nil, err
+				}
+				if found {
+					return &ResolveCheckResponse{
+						Allowed: true,
+						ResolutionMetadata: &ResolveCheckResponseMetadata{
+							// It probably is +1.  However, we have no control on whether
+							// the userset member lookup for the first level has started DS query.
+							DatastoreQueryCount: req.GetRequestMetadata().DatastoreQueryCount + 2,
+						},
+					}, nil, nil, nil
+				}
+			}
+		}
+	}
+	if usersetFromObject.Size() == 0 {
+		return &ResolveCheckResponse{
+			Allowed: false,
+			ResolutionMetadata: &ResolveCheckResponseMetadata{
+				// It probably is +1.  However, we have no control on whether
+				// the userset member lookup for the first level has started DS query.
+				DatastoreQueryCount: req.GetRequestMetadata().DatastoreQueryCount + 2,
+			},
+		}, nil, nil, nil
+	}
+	return nil, usersetFromUser, usersetFromObject, nil
+}
+
+func nestedUsersetFastpath(ctx context.Context,
+	typesys *typesystem.TypeSystem,
+	ds storage.RelationshipTupleReader,
+	req *ResolveCheckRequest,
+	maxConcurrentReads int) (*ResolveCheckResponse, error) {
+	ctx, span := tracer.Start(ctx, "nestedUsersetFastpath")
+	defer span.End()
+
+	reqTupleKey := req.GetTupleKey()
+	objectType := tuple.GetType(reqTupleKey.GetObject())
+	relation := reqTupleKey.GetRelation()
+
+	userToUsersetMessageChan := streamedLookupUsersetForUser(ctx, typesys, ds, req, maxConcurrentReads)
```

Comment:
Within `streamedLookupUsersetForUser` and `streamedLookupUsersetForObject` the only way for you to realize when the goroutines are done is when the channels are closed. In every path where you shortcircuit, e.g. in line 1452, how are we ensuring that we are not forgetting about those two goroutines?

Reply:
Good callout - fixed.